text/3617-precise-capturing.md

```diff
@@ -0,0 +1,668 @@
+
+Two, each keyword takes from the space of names that users have available to them, and it increases the number of keywords with which users must be familiar (e.g. so as to not inadvertently trip over when choosing a name).  That is, each keyword has a cost.  If an existing keyword can reasonably be used in more places, then we get more benefit for that cost.  In this case, `use` is probably a strong enough choice that paying the cost for a new keyword doesn't seem worth it.
+
+Three, `captures` would be a somewhat long keyword, especially when we consider how we might scale the use of this syntax to other places such as closure-like blocks.  We don't want people to feel punished for being explicit about the generics that they capture, and we don't want them to do other worse things (such as overcapturing where they should not) just to avoid visual bloat in their code, so if we can be more concise here, that seems like a win.
+
+### `impl move<'t, T> Trait`
+
+We could use the existing `move` keyword, however the word "move" is semantically worse.  In Rust, we already *use* generic parameters in types, but we don't *move* any generic parameters.  We move only *values*, so this could be confusing.  The word "use" is better.
+
+### `impl k#via<'t, T> Trait`
+
+We could use a new short keyword such as `via`.  This has the number 1 and 2 drawbacks of `k#captures` mentioned above.  As with `move`, it also seems a semantically worse word.  With `use<..>`, we can explain that it means the opaque type *uses* the listed generic parameters.  In contrast, it's not clear how we could explain the word "via" in this context.
+
+### Using parenthesis or square brackets
+
+We could say `use('t, T)` or `use['t, T]`.  However, in Rust today, generic parameters always fall within angle brackets, even when being applied to a type.  Doing something different here could feel inconsistent and doesn't seem warranted.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Opting out of captures
+
+There will plausibly be cases where we want to capture many generic parameters and not capture only smaller number.  It could be convenient if there were a way to express this without listing out all of the in-scope type parameters except the ones not being captured.
+
+The way we would approach this with the `use<..>` syntax is to add some syntax that means "fill in all in-scope generic parameters", then add syntax to remove certain generic parameters from the list.  E.g.:
+
+```rust
+fn foo<'a, A, B, C, D>(
+    _: &'a A, b: B, c: C, d: D,
+) -> impl use<.., !'a, !A> Sized {}
+//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+//   ^ Captures `B`, `C`, and `D` but not `'a` or `A`.
+```
+
+Here, the `..` means to include all in-scope generic parameters and `!` means to exclude a particular generic parameter even if previously included.
```

Comment:
It might be nice to allow `use<..>` from the start as an explicit way to capture all in-scope generic parameters. Using this short-hand together with explicit explicit captures or excluded captures should still remain future work.

Especially in code where precise captures would be frequently used, it might be preferable to always be explicit, even in cases where the implicit capture-all is what we want. Similar to how the syntax is used with struct initialisation, using `use<..>` could mean that we know that we will always want to capture all generics, even if more are added, whereas manually enumerating each one might mean that we want to explicitly have to choose with every newly added generic.

Furthermore, since the stabilisation strategy suggests that at first only capturing all parameters may be supported, supporting `use<..>` from the start would make this shorter to type.

Reply:
An alternative meaning for `use<..>` is "capture all APITs".